pmlc/dialect/comp/ir/interfaces.td

```diff
@@ -0,0 +1,80 @@
+// Copyright 2020 Intel Corporation
+#ifndef __PMLC_COMP_INTERFACES__
+#define __PMLC_COMP_INTERFACES__
+
+include "mlir/IR/OpBase.td"
+
+class COMP_OpInterface<string name> : OpInterface<name>;
+
+def COMP_ExecEnvOpInterface : COMP_OpInterface<"ExecEnvOpInterface"> {
+  let description = [{
+    Ties operation to specific execution environment.
+  }];
+
+  let methods = [
+    InterfaceMethod<"", "::mlir::Value", "getExecEnv", (ins), [{}], [{
+      ConcreteOp op = ::mlir::cast<ConcreteOp>(this->getOperation());
+      return op.execEnv();
+    }]>
+  ];
+}
+
+def COMP_ScheduleOpInterface : COMP_OpInterface<"ScheduleOpInterface"> {
```

Comment:
I'm also not sure the concept of "do something asynchronously" should be represented at this level at all -- what's synchronous and what's asynchronous may vary from API to API.  IMHO, it might be better to simply represent the dependencies between operations at this level, and let the lowering to target API consider how to handle asynchronicity.

Reply:
I have added dialect description addressing this concern. I agree that there should be no requirement for asynchronous execution here and I have tried to avoid it. All the operations in `comp` dialect could be lowered to synchronous API in which case events and explicit dependencies can be ignored/removed.
This is ensured by the fact that "schedule" operations don't guarantee time after scheduling that operations will finish executing (it can be immediately)  and that all events are SSA values so operations can only depend on previously scheduled operations.